File: Framework/CSharp/Framework/Framework/Diagnostics/SmartStopwatch.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Diagnostics;

namespace Smartkernel.Framework.Diagnostics
{
    /// <summary>
    /// 计时器
    /// </summary>
    public static class SmartStopwatch
    {
        /// <summary>
        /// 运行
        /// </summary>
        /// <param name="action">行为</param>
        /// <returns>结果</returns>
        public static TimeSpan Execute(Action action)
        {
            var stopwatch = new Stopwatch();
            stopwatch.Start();

            action();

            stopwatch.Stop();
            return stopwatch.Elapsed;
        }
    }
}

File: Framework/Php/framework/framework/diagnostics/smartstopwatch.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework\Diagnostics {
    /*
     * 计时器
     */

    class SmartStopwatch {

        private $startTime = 0;
        private $stopTime = 0;

        /*
         * 开始
         * 
         */

        public function start() {
            list($usec, $sec) = explode(" ", microtime());
            $this->startTime = (float) $usec + (float) $sec;
            $this->stopTime = (float) $usec + (float) $sec;
        }

        /*
         * 结束
         */

        public function stop() {
            list($usec, $sec) = explode(" ", microtime());
            $this->stopTime = (float) $usec + (float) $sec;
        }

        /*
         * 重启
         */

        public function restart() {
            $this->start();
        }

        /*
         * 持续时间
         * 
         * @return float，结果
         * 
         */

        public function elapsed() {
            return ($this->stopTime - $this->startTime) * 1000;
        }

        /*
         * 运行
         * 
         * @param function $action，行为
         * @return float，结果
         * 
         */

        public static function execute($action) {
            $stopwatch = new SmartStopwatch();
            $stopwatch->start();

            $action();

            $stopwatch->stop();
            return $stopwatch->elapsed();
        }

    }

}